pds_doi_service/api/test/test_dois_controller.py

```diff
@@ -277,6 +275,212 @@ def test_post_dois_invalid_requests(self):
+        )
+
+        # Recreate a DoiRecord from the response JSON and examine the
+        # fields
+        release_record = DoiRecord.from_dict(release_response.json[0])
+
+        self.assertEqual(release_record.submitter, 'eng-submitter@jpl.nasa.gov')
+        self.assertEqual(release_record.lid, 'urn:nasa:pds:insight_cameras')
+        self.assertEqual(release_record.vid, '1.1')
+        self.assertEqual(release_record.status, 'Released')
+        self.assertEqual(release_record.doi, '10.17189/21734')
+
+        # Record field should match what we provided via patch method
+        self.assertEqual(release_record.record, self.release_action_run_patch())
+
+    def release_action_run_w_error_patch(self, **kwargs):
+        """
+        Patch for DOICoreActionRelease.run()
+
+        Returns body of an OSTI XML label corresponding to errors returned
+        from OSTI.
+        """
+        draft_record_file = join(
+            TestDoisController.test_data_dir, 'error_osti_record.xml')
+        with open(draft_record_file, 'r') as infile:
+            return infile.read()
+
+    @patch.object(
+        pds_doi_service.api.controllers.dois_controller.DOICoreActionRelease,
+        'run', release_action_run_w_error_patch)
+    @patch.object(
+        pds_doi_service.api.controllers.dois_controller.DOICoreActionList,
+        'run', list_action_run_patch)
+    def test_post_release_w_errors(self):
+        """
+        Test the release endpoint where errors are received back from the
+        release action.
+        """
+        query_string = [('force', False),
+                        ('db_name', self.temp_db)]
+
+        error_response = self.client.open(
+            '/PDS_APIs/pds_doi_api/0.1/dois/{lid}/release'
+            .format(lid='urn:nasa:pds:insight_cameras::1.1'),
+            method='POST',
+            query_string=query_string
+        )
+
+        self.assert400(
+            error_response,
+            'Response body is : ' + error_response.data.decode('utf-8')
+        )
+
+        # Check the error response and make sure it contains all the errors
+        # provided from the original XML
+        errors = error_response.json['errors']
+
+        self.assertEqual(errors[0]['name'], 'WarningDOIException')
+        self.assertIn('Title is required', errors[0]['message'])
+        self.assertIn('A publication date is required', errors[0]['message'])
+        self.assertIn('A site URL is required', errors[0]['message'])
+        self.assertIn('A product type is required', errors[0]['message'])
+        self.assertIn('A specific product type is required for non-dataset types',
+                      errors[0]['message'])
+
+    def list_action_run_patch_missing(self, **kwargs):
+        """
+        Patch for DOICoreActionList.run()
+
+        Returns a result corresponding to an unsuccessful search.
+        """
+        return '[]'
+
+    @patch.object(
```

Comment:
Almost, the `patch.object` decorator lets you mock individual fields or functions from class, rather than the whole class itself. In this case, the `run` methods from the action objects is being mocked to return a canned OSTI response XML file.